File: mainstream-android/app/src/main/java/com/mainstream/app/data/model/Trophy.kt
package com.mainstream.app.data.model

import com.google.gson.annotations.SerializedName

data class Trophy(
    val id: Long,
    val code: String,
    val name: String,
    val description: String,
    val type: TrophyType,
    val category: TrophyCategory,
    val iconUrl: String? = null,
    val criteriaValue: Double? = null,
    val isActive: Boolean,
    val displayOrder: Int? = null,
    val latitude: Double? = null,
    val longitude: Double? = null,
    val collectionRadiusMeters: Double? = null,
    val validFrom: String? = null,
    val validUntil: String? = null,
    val imageUrl: String? = null,
    val criteriaConfig: String? = null,
    val checkScope: CheckScope? = null,
    val createdAt: String,
    val updatedAt: String
)

data class UserTrophy(
    val id: Long,
    val userId: Long,
    val userName: String,
    val trophy: Trophy,
    val activityId: Long? = null,
    val earnedAt: String,
    val metadata: String? = null
)

data class TrophyProgress(
    val trophyId: Long,
    val currentValue: Double,
    val targetValue: Double,
    val percentage: Double,
    val isComplete: Boolean,
    val isEarned: Boolean
)

data class TrophyWithProgress(
    val trophy: Trophy,
    val isEarned: Boolean,
    val earnedAt: String? = null,
    val progress: TrophyProgress? = null
)

enum class TrophyType {
    @SerializedName("DISTANCE_MILESTONE") DISTANCE_MILESTONE,
    @SerializedName("STREAK") STREAK,
    @SerializedName("ROUTE_COMPLETION") ROUTE_COMPLETION,
    @SerializedName("CONSISTENCY") CONSISTENCY,
    @SerializedName("TIME_BASED") TIME_BASED,
    @SerializedName("EXPLORER") EXPLORER,
    @SerializedName("LOCATION_BASED") LOCATION_BASED,
    @SerializedName("SPECIAL") SPECIAL
}

enum class TrophyCategory {
    @SerializedName("BEGINNER") BEGINNER,
    @SerializedName("INTERMEDIATE") INTERMEDIATE,
    @SerializedName("ADVANCED") ADVANCED,
    @SerializedName("ELITE") ELITE,
    @SerializedName("SPECIAL") SPECIAL
}

enum class CheckScope {
    @SerializedName("SINGLE_ACTIVITY") SINGLE_ACTIVITY,
    @SerializedName("TOTAL") TOTAL,
    @SerializedName("TIME_RANGE") TIME_RANGE,
    @SerializedName("COUNT") COUNT
}

fun TrophyCategory.getDisplayName(): String = when (this) {
    TrophyCategory.BEGINNER -> "Anfänger"
    TrophyCategory.INTERMEDIATE -> "Fortgeschritten"
    TrophyCategory.ADVANCED -> "Erfahren"
    TrophyCategory.ELITE -> "Elite"
    TrophyCategory.SPECIAL -> "Spezial"
}

fun TrophyCategory.getColor(): Long = when (this) {
    TrophyCategory.BEGINNER -> 0xFF4CAF50 // Green
    TrophyCategory.INTERMEDIATE -> 0xFF2196F3 // Blue
    TrophyCategory.ADVANCED -> 0xFF9C27B0 // Purple
    TrophyCategory.ELITE -> 0xFFFF9800 // Orange
    TrophyCategory.SPECIAL -> 0xFFF44336 // Red
}

File: mainstream-ios/Mainstream/Models/Trophy.swift
import Foundation
import SwiftUI

// MARK: - Trophy Model
struct Trophy: Codable, Identifiable {
    let id: Int
    let code: String
    let name: String
    let description: String
    let type: TrophyType
    let category: TrophyCategory
    let iconUrl: String?
    let criteriaValue: Double?
    let isActive: Bool
    let displayOrder: Int?

    // Location-based trophy fields
    let latitude: Double?
    let longitude: Double?
    let collectionRadiusMeters: Double?
    let validFrom: String?
    let validUntil: String?
    let imageUrl: String?

    // Generic configurable trophy criteria
    let criteriaConfig: String?
    let checkScope: CheckScope?

    let createdAt: String
    let updatedAt: String
}

// MARK: - Trophy Enums
enum TrophyType: String, Codable {
    case distanceMilestone = "DISTANCE_MILESTONE"
    case streak = "STREAK"
    case routeCompletion = "ROUTE_COMPLETION"
    case consistency = "CONSISTENCY"
    case timeBased = "TIME_BASED"
    case explorer = "EXPLORER"
    case locationBased = "LOCATION_BASED"
    case special = "SPECIAL"

    var displayName: String {
        switch self {
        case .distanceMilestone: return "Distanz-Meilenstein"
        case .streak: return "Serie"
        case .routeCompletion: return "Streckenabschluss"
        case .consistency: return "Beständigkeit"
        case .timeBased: return "Zeitbasiert"
        case .explorer: return "Entdecker"
        case .locationBased: return "Standortbasiert"
        case .special: return "Spezial"
        }
    }
}

enum TrophyCategory: String, Codable {
    case beginner = "BEGINNER"
    case intermediate = "INTERMEDIATE"
    case advanced = "ADVANCED"
    case elite = "ELITE"
    case special = "SPECIAL"

    var displayName: String {
        switch self {
        case .beginner: return "Anfänger"
        case .intermediate: return "Fortgeschritten"
        case .advanced: return "Erweitert"
        case .elite: return "Elite"
        case .special: return "Spezial"
        }
    }

    var color: Color {
        switch self {
        case .beginner: return Color.green
        case .intermediate: return Color.blue
        case .advanced: return Color.orange
        case .elite: return Color.purple
        case .special: return Color.red
        }
    }
}

enum CheckScope: String, Codable {
    case singleActivity = "SINGLE_ACTIVITY"
    case total = "TOTAL"
    case timeRange = "TIME_RANGE"
    case count = "COUNT"
}

// MARK: - User Trophy
struct UserTrophy: Codable, Identifiable {
    let id: Int
    let userId: Int
    let userName: String
    let trophy: Trophy
    let activityId: Int?
    let earnedAt: String
    let metadata: String?

    var earnedDate: Date? {
        ISO8601DateFormatter().date(from: earnedAt)
    }
}

// MARK: - Trophy With Progress
struct TrophyWithProgress: Codable, Identifiable {
    let id: Int
    let code: String
    let name: String
    let description: String
    let type: TrophyType
    let category: TrophyCategory
    let iconUrl: String?
    let criteriaValue: Double?
    let isActive: Bool
    let displayOrder: Int?
    let createdAt: String
    let updatedAt: String

    let isEarned: Bool
    let earnedAt: String?
    let progress: Double?
    let progressMax: Double?

    var progressPercentage: Double {
        guard let progress = progress, let progressMax = progressMax, progressMax > 0 else {
            return 0
        }
        return min(progress / progressMax * 100, 100)
    }
}
